app/docker/helpers/containerHelper.js

```diff
@@ -54,5 +54,170 @@ angular.module('portainer.docker')
+  };
+
+  helper.preparePortBindings = function(portBindings) {
+    var bindings = {};
+    portBindings.forEach(function (portBinding) {
+      if (portBinding.containerPort) {
+        var hostPort = portBinding.hostPort;
+        var containerPortRange = helper.parsePortRange(portBinding.containerPort);
+        if (!containerPortRange || containerPortRange.length !== 2) {
+          throw new Error('Invalid port specification: ' + portBinding.containerPort);
+        }
+
+        var startPort = containerPortRange[0];
+        var endPort = containerPortRange[1];
+        var hostIp = undefined;
+        var startHostPort = 0;
+        var endHostPort = 0;
+        if (hostPort) {
+          if (hostPort.indexOf(':') > -1) {
+            var hostAndPort = hostPort.split(':');
+            hostIp = hostAndPort[0];
+            hostPort = hostAndPort[1];
+          }
+
+          var hostPortRange = helper.parsePortRange(hostPort);
+          if (!hostPortRange || hostPortRange.length !== 2) {
+            throw new Error('Invalid port specification: ' + hostPort);
+          }
+
+          startHostPort = hostPortRange[0];
+          endHostPort = hostPortRange[1];
+          if ((endPort - startPort) !== (endHostPort - startHostPort)) {
+            if (endPort !== startPort) {
+              throw new Error('Invalid port specification: ' + hostPort);
+            }
+          }
+        }
+
+        for (var i = 0; i <= (endPort - startPort); i++) {
+          var containerPort = (startPort + i).toString();
+          if (startHostPort > 0) {
+            hostPort = (startHostPort + i).toString();
+          }
+          if (startPort === endPort && startHostPort !== endHostPort) {
+            hostPort += '-' + endHostPort.toString();
+          }
+
+          var key = containerPort + '/' + portBinding.protocol;
+          bindings[key] = [{ HostIp: hostIp, HostPort: hostPort }];
+        }
+      }
+    });
+    return bindings;
+  };
+
+  helper.createPortRange = function(portRange, port) {
+    var hostIp = null;
+    var colonIndex = portRange.indexOf(':');
+    if (colonIndex >= 0) {
+      hostIp = portRange.substr(0, colonIndex);
+      portRange = portRange.substr(colonIndex + 1);
+    }
+
+    var startPort;
+    var endPort;
+    port = (typeof port === 'number' ? port : parseInt(port));
+    portRange = helper.parsePortRange(portRange);
+    if (!portRange || portRange.length !== 2) {
+      var portSplit = portRange.split('-');
+      startPort = parseInt(portSplit[0]);
+      endPort = startPort;
+    } else {
+      startPort = portRange[0];
+      endPort = portRange[1];
+    }
+
+    if (port < startPort) {
+      startPort = port;
+    }
+    if (port > endPort) {
+      endPort = port;
+    }
+
+    if (hostIp) {
+      return hostIp + ':' + startPort + '-' + endPort;
+    } else {
+      return startPort + '-' + endPort;
+    }
+  };
+
+  helper.sortAndCombinePorts = function(portBindings) {
```

Comment:
This function needs a refactor as there is no differenciation or verification
* between TCP and UDP when building the ranges.
* between the bound host IP and the one of the range that is currently beeing constructed.

Need to do :
* Keys should be splitted between TCP and UDP, then sorted separatly.
* For each merge by common host IP
* After these 2 steps, building the ranges will be easier.

Reply:
You are absolutely right. I totally missed that. It should be resolved now.